EPUB cover too large for the database

Normal text epub file is working but we are unable to open **audio, video multimedia** Epub file.
Epub 3 supports multimedia but it does not work for me with this code.

We are using `v2.2.0`
We tried with given file https://github.com/IDPF/epub3-samples/releases/download/20170606/cc-shared-culture.epub it works for me but below file contain video and when i open file, app is closed.  
https://kstyles.app/wp-content/uploads/2020/08/11.epub

`` FATAL EXCEPTION: main
    Process: org.readium.r2reader, PID: 32663
    java.lang.RuntimeException: Unable to start activity ComponentInfo{org.readium.r2reader/org.readium.r2.testapp.epub.EpubActivity}: android.database.sqlite.SQLiteBlobTooBigException: Row too big to fit into CursorWindow requiredPos=0, totalRows=1
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2946)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3081)
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:78)
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108)
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1831)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:201)
        at android.app.ActivityThread.main(ActivityThread.java:6810)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:547)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:873)
     Caused by: android.database.sqlite.SQLiteBlobTooBigException: Row too big to fit into CursorWindow requiredPos=0, totalRows=1
        at android.database.sqlite.SQLiteConnection.nativeExecuteForCursorWindow(Native Method)
        at android.database.sqlite.SQLiteConnection.executeForCursorWindow(SQLiteConnection.java:859)
        at android.database.sqlite.SQLiteSession.executeForCursorWindow(SQLiteSession.java:836)
        at android.database.sqlite.SQLiteQuery.fillWindow(SQLiteQuery.java:62)
        at android.database.sqlite.SQLiteCursor.fillWindow(SQLiteCursor.java:151)
        at android.database.sqlite.SQLiteCursor.getCount(SQLiteCursor.java:139)
        at org.jetbrains.anko.db.SqlParsersKt.parseList(SqlParsers.kt:79)
        at org.readium.r2.testapp.db.BOOKS$currentLocator$1$1.invoke(BooksDatabase.kt:260)
        at org.readium.r2.testapp.db.BOOKS$currentLocator$1$1.invoke(BooksDatabase.kt:200)
        at org.jetbrains.anko.db.SelectQueryBuilder.exec(SelectQueryBuilder.kt:44)
        at org.readium.r2.testapp.db.BOOKS$currentLocator$1.invoke(BooksDatabase.kt:259)
        at org.readium.r2.testapp.db.BOOKS$currentLocator$1.invoke(BooksDatabase.kt:200)
        at org.jetbrains.anko.db.ManagedSQLiteOpenHelper.use(Database.kt:177)
        at org.readium.r2.testapp.db.BOOKS.currentLocator(BooksDatabase.kt:256)
        at org.readium.r2.testapp.epub.EpubActivity.onCreate(EpubActivity.kt:238)
        at android.app.Activity.performCreate(Activity.java:7224)
        at android.app.Activity.performCreate(Activity.java:7213)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1272)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2926)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3081) 
        at android.app.servertransaction.LaunchActivityItem.execute(LaunchActivityItem.java:78) 
        at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:108) 
        at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:68) 
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1831) 
        at android.os.Handler.dispatchMessage(Handler.java:106) 
        at android.os.Looper.loop(Looper.java:201) 
        at android.app.ActivityThread.main(ActivityThread.java:6810) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:547) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:873) 
``

--- a/r2-testapp/src/main/java/org/readium/r2/testapp/db/BooksDatabase.kt
+++ b/r2-testapp/src/main/java/org/readium/r2/testapp/db/BooksDatabase.kt
@@ -14,15 +14,22 @@ import android.content.ContentValues
 import android.content.Context
 import android.database.sqlite.SQLiteDatabase
 import android.database.sqlite.SQLiteException
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
 import android.net.Uri
 import android.os.Build
+import android.util.Size
+import kotlinx.coroutines.runBlocking
 import org.jetbrains.anko.db.*
 import org.joda.time.DateTime
 import org.json.JSONObject
+import org.readium.r2.shared.extensions.toPng
+import org.readium.r2.shared.extensions.tryOrNull
 import org.readium.r2.shared.publication.Locator
 import org.readium.r2.shared.publication.Publication
 import java.net.URI
 import java.nio.file.Paths
+import kotlin.math.min
 
 /**
  * Global Parameters
@@ -208,13 +215,22 @@ class BOOKS(private var database: BooksDatabaseOpenHelper) {
     fun insert(book: Book, allowDuplicates: Boolean): Long? {
         val exists = has(book)
         if (exists.isEmpty() || allowDuplicates) {
+            // Makes sure the cover is not too large, to prevent SQLiteBlobTooBigException.
+            val cover = book.cover
+                ?.let {
+                    tryOrNull {
+                        BitmapFactory.decodeByteArray(it, 0, it.size)
+                            ?.scaleToFit(Size(400, 400))
+                    }
+                }
+
             return database.use {
                 return@use insert(BOOKSTable.NAME,
                         BOOKSTable.TITLE to book.title,
                         BOOKSTable.AUTHOR to book.author,
                         BOOKSTable.HREF to book.href,
                         BOOKSTable.IDENTIFIER to book.identifier,
-                        BOOKSTable.COVER to book.cover,
+                        BOOKSTable.COVER to runBlocking { cover?.toPng() },
                         BOOKSTable.EXTENSION to book.ext,
                         BOOKSTable.CREATION to book.creation)
             }
@@ -337,4 +353,19 @@ class BOOKS(private var database: BooksDatabaseOpenHelper) {
             return Book(id, creation as Long, href, title, author, identifier, cover, progression, ext)
         }
     }
+}
+
+private fun Bitmap.scaleToFit(maxSize: Size): Bitmap {
+    if (width <= maxSize.width && height <= maxSize.height)
+        return this
+
+    val ratio = min(
+        maxSize.width / width.toFloat(),
+        maxSize.height / height.toFloat()
+    )
+
+    val newWidth = (ratio * width).toInt()
+    val newHeight = (ratio * height).toInt()
+
+    return Bitmap.createScaledBitmap(this, newWidth, newHeight, true)
 }

